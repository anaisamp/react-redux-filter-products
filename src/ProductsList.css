.categoriesWrapper {
    display: flex;
    flex-wrap: wrap;
    margin-bottom: 8px;
}

.categoryTitle {
    padding: 6px 8px 6px 0px;
    cursor: pointer;
    font-size: 14px;
}

.categoryTitle.selected {
    font-weight: 700;
    text-decoration: underline;
}

.productList {
    margin-top: 8px;
}

.productWrapper {
    padding: 3px 8px 3px 0px;
}

.productTitle {
    margin: 0;
    padding: 3px 0;
    cursor: pointer;
    font-size: 12px;
}

.productTitle.selected {
    font-weight: 700;
}

.productDescription {
    margin: 0;
    font-size: 12px;
    font-style: italic;
}

.hide {
    display: none;
}

.show {
    display: block;
}