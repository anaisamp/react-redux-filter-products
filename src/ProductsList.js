import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import classnames from 'classnames';
import { fetchCategoriesAndProducts } from './actions/categories';
import { fetchProducts, toogleProductDescriptionVisibility, fetchProductsFilter } from './actions/products';
import './ProductsList.css';

  class ProductsList extends React.Component {

    handleClickSelectCategory = (categoryId) => {
        this.props.fetchProducts(categoryId);
    }
   
    handleClickExpandDescription = (index) => {
        this.props.toogleProductDescriptionVisibility(index);
    }

    handleSearch = (searchValue) => {
        this.props.fetchProductsFilter(searchValue);
    }

    componentDidMount() {
        if(this.props.categories.length === 0) {
            this.props.fetchCategoriesAndProducts();
        }
    }

    render() {
        return (
            <div>
                <div className="categoriesWrapper">
                { 
                    this.props.categories.map(c => (
                        <div
                            key={ c.id } 
                            className={ 
                                classnames({
                                    'categoryTitle': true,
                                    'selected': this.props.selectedCategory === c.id,
                                }) } 
                            onClick={ () => this.handleClickSelectCategory(c.id) } 
                            selected={this.props.selectedCategory === c.id}
                        >
                            { c.title }
                        </div>
                    ))
                }
                </div>
                <input type="search" size="45" 
                    value= { this.props.searchValue }
                    onInput={ (e) => this.handleSearch(e.target.value) }
                />
                <div className="productList">
                {
                    this.props.visibleProducts.map((p, i) => (  
                        <div key={p.id}  className="productWrapper">
                            <p onClick={ () => this.handleClickExpandDescription(i)}
                                className={ `productTitle ${this.props.toogleProductDescription[i] ? 'selected' : ''}`}>
                                { p.title }
                            </p>
                            <p className={  `productDescription ${this.props.toogleProductDescription[i] ? 'show' : 'hide'}`}> 
                                { p.description } 
                            </p>
                        </div>
                    ))
                }
                </div>
        </div>);
    }
  }

  const mapStateToProps = (state, ownProps) => ({
    categories: state.categories.categories,
    isFetchingCategories: state.categories.isFetchingCategories,
    errorFetchingCategories: state.categories.errorFetchingCategories,
    products: state.products.products,
    visibleProducts: state.products.visibleProducts,
    isFetchingProducts: state.products.isFetchingProducts,
    errorFetchingProducts: state.products.errorFetchingProducts,
    selectedCategory: state.products.selectedCategory,
    toogleProductDescription: state.products.toogleProductDescription,
    searchValue: state.products.searchValue,
  });
  
  const mapDispatchToProps = dispatch => bindActionCreators({
    fetchProducts,
    fetchProductsFilter,
    fetchCategoriesAndProducts,
    toogleProductDescriptionVisibility,
  }, dispatch);
  
  export default connect(
    mapStateToProps,
    mapDispatchToProps
  )(ProductsList);