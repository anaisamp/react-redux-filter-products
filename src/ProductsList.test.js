import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import store from './store';
import ProductsList from './ProductsList';

Enzyme.configure({ adapter: new Adapter() })

describe('ProductsList', () => {
    it('maps store props and actions correctly with initial state', () => {
        const wrapper = shallow(
            <ProductsList store={store} />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
