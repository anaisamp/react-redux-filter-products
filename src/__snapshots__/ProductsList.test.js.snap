// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsList maps store props and actions correctly with initial state 1`] = `
<ProductsList
  categories={Array []}
  errorFetchingCategories={false}
  errorFetchingProducts={false}
  fetchCategoriesAndProducts={[Function]}
  fetchProducts={[Function]}
  fetchProductsFilter={[Function]}
  isFetchingCategories={false}
  isFetchingProducts={false}
  products={Array []}
  searchValue=""
  selectedCategory={null}
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "unsubscribe": [Function],
    }
  }
  toogleProductDescription={Array []}
  toogleProductDescriptionVisibility={[Function]}
  visibleProducts={Array []}
/>
`;
